plugins/inputs/huawei_routers_telemetry/huawei_routers_telemetry.go

```diff
@@ -0,0 +1,219 @@
+	buf := make([]byte, 64*1024) // 64kb - maximum size of IP packet
+	for {
+		n, _, err := h.ReadFrom(buf)
+		if err != nil {
+			h.Log.Error("Unable to read buffer: %s", err.Error())
+			break
+		}
+
+		body, err := h.decoder.Decode(buf[:n])
+		if err != nil {
+			h.Log.Errorf("Unable to decode incoming packet: %s", err.Error())
+			continue
+		}
+		// Telemetry parsing over packet payload
+		grouper, err := HuaweiTelemetryDecoder(body)
+		if err != nil {
+			h.Log.Errorf("Unable to decode telemetry information: %s", err.Error())
+			break
+		}
+		for _, metric := range grouper.Metrics() {
+			h.AddMetric(metric)
+		}
+
+		if err != nil {
+			h.Log.Errorf("Unable to parse incoming packet: %s", err.Error())
+		}
+	}
+}
+
+type HuaweiRoutersTelemetry struct {
+	ServicePort     string        `toml:"service_port"`
+	ReadBufferSize  internal.Size `toml:"read_buffer_size"`
+	ContentEncoding string        `toml:"content_encoding"`
+	wg              sync.WaitGroup
+
+	Log telegraf.Logger `toml:"-"`
+
+	telegraf.Accumulator
+	io.Closer
+	decoder internal.ContentDecoder
+}
+
+func (h *HuaweiRoutersTelemetry) Description() string {
+	return "Huawei Telemetry UDP model input Plugin"
+}
+
+func (h *HuaweiRoutersTelemetry) SampleConfig() string {
+	return `
+  ## UDP Service Port to capture Telemetry
+  # service_port = "8080"
+
+`
+}
+
+func (h *HuaweiRoutersTelemetry) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (h *HuaweiRoutersTelemetry) Start(acc telegraf.Accumulator) error {
+	h.Accumulator = acc
+
+	var err error
+	h.decoder, err = internal.NewContentDecoder(h.ContentEncoding)
+	if err != nil {
+		return err
+	}
+
+	pc, err := udpListen("udp", ":"+h.ServicePort)
+	if err != nil {
+		return err
+	}
+
+	if h.ReadBufferSize.Size > 0 {
+		if srb, ok := pc.(setReadBufferer); ok {
+			srb.SetReadBuffer(int(h.ReadBufferSize.Size))
+		} else {
+			h.Log.Warnf("Unable to set read buffer on a %s socket", "udp")
+		}
+	}
+
+	h.Log.Infof("Listening on %s://%s", "udp", pc.LocalAddr())
+
+	psl := &packetSocketListener{
+		PacketConn:             pc,
+		HuaweiRoutersTelemetry: h,
+	}
+
+	h.Closer = psl
+	h.wg = sync.WaitGroup{}
+	h.wg.Add(1)
+	go func() {
+		defer h.wg.Done()
+		psl.listen()
+	}()
+	return nil
+}
+
+func udpListen(network string, address string) (net.PacketConn, error) {
+	switch network {
+	case "udp", "udp4", "udp6":
+		var addr *net.UDPAddr
+		var err error
+		var ifi *net.Interface
+		if spl := strings.SplitN(address, "%", 2); len(spl) == 2 {
+			address = spl[0]
+			ifi, err = net.InterfaceByName(spl[1])
+			if err != nil {
+				return nil, err
+			}
+		}
+		addr, err = net.ResolveUDPAddr(network, address)
+		if err != nil {
+			return nil, err
+		}
+		if addr.IP.IsMulticast() {
+			return net.ListenMulticastUDP(network, ifi, addr)
+		}
+		return net.ListenUDP(network, addr)
+	}
+	return net.ListenPacket(network, address)
+}
+
+func (h *HuaweiRoutersTelemetry) Stop() {
+	if h.Closer != nil {
+		h.Close()
+		h.Closer = nil
+	}
```

Comment:
Please add the connection as a field to the `HuaweiRoutersTelemetry` struct and do
```suggestion
	if h.connection != nil {
		h.connection.Close()
	}
```

Reply:
done